oeps/oep-0049-django-app-patterns.rst

```diff
@@ -0,0 +1,211 @@
+
+
+Not exposing an app's data structures can be tricky because it's very easy to expose them without meaning to. Therefore there are a couple common strategies we employ.
+
+1. When importing internal app code to be used in the ``api.py`` file, prefix it with an underscore so it's clear it's for internal use only.
+
+2. Create a ``data.py`` file to house simple data objects that can be passed from your app's function to the calling app. By creating these objects, we can avoid both passing Django model objects or querysets directly and having to serialize data. Other apps may import data classes from ``data.py`` in additional to functionality from ``api.py``. See data.py_ for more details.
+
+Using Django's Paginator class can help keep cross-app retrievals performant without passing Querysets around.
+
+For example:
+
+.. code-block:: python
+
+  from django.conf.settings import UNSUPPORTED_PROGRAM_UUIDS
+
+  from .data import ProgramData
+  from .models_api import get_programs as _get_programs
+
+  def get_supported_programs(page_size=None, page=None):
+      """
+      Gets all programs that aren't in UNSUPPORTED_PROGRAM_UUIDS settings
+
+      Returns a page of results if page_size is specified
+      """
+      # _get_programs() returns a queryset
+      q_supported_programs = _get_programs().exclude(
+          uuid__in=UNSUPPORTED_PROGRAM_UUIDS
+      )
+
+      if page_size:
+          # passing a queryset to a paginator allows it to stay performant on large tables
+          program_paginator = Paginator(q_supported_programs, page_size)
+          # get_page returns the first page if page is None
+          supported_programs = program_paginator.get_page(page)
+      else:
+          supported_programs = q_supported_programs
+
+      return [
+          ProgramData(
+              uuid=program.uuid,
+              title=program.title,
+              status=program.status
+          )
+          for program in supported_programs
+      ]
+
+.. _data.py:
+data.py
++++++++
+This file should include the public data structures for the app that can be passed between apps without exposing internal features. These should be used instead of sending Django model objects or querysets to apps that call the functions in ``api.py``. This file should not import anything other than stdlib modules, so that it may be imported by any other app without issue. These data objects should be simple objects with all business logic handled by ``api.py``. They may however perform simple validation, as long as it is self-contained (doesn't reach out to database, network, or any code outside of the class)
+
+For example:
+
+.. code-block:: python
+
+
+  from dataclasses import dataclass
+  from enum import Enum
+
+  import attr
+
+  def ProgramStatus(Enum):
+      ACTIVE = "active"
+      RETIRED = "retired"
+
+  @attr.attrs(frozen=True)
+  class ProgramData:
+      uuid: str = attr.attrib(
+          validator=[attr.validators.instance_of(str)]
```

Comment:
Is it too much to hope for mypy running, so we don't have to do runtime checking of this?